# Error types
Add error types and expand `Response` left options to include them for all domain entities/aggregates.

I want to return an error object with a message and an optional object containing data about the error.

Looking more closely at Stemmler's `Guard` behavior, I see a need to make some changes to be more like his definition.

## Make Either clearer
[x] Rename types on `Either` to make them easier to read. L -> LeftType; R -> RightType

**COMMIT: 3.1.7.1 - REFACTOR: Rename types on Either to be clearer**

## Define base Error classes
What needs to go into these classes?

I want to include a result code of some kind to better support analysis. Would that be in errors or in `Result`?
* If in `Result` does it have any value for anything that isn't a failure?
  * Probably not. So could be defined for the base error and returned in the error value.
* Are error messages the same for every instance of the same error?
  * Yes, if errors are properly specific, so make them specific.
* How do I want to group errors?
  * Stemmler does it by use case (he generates all errors from use cases; failures in entities are `Result`s).
  * Some errors seem to apply to several use case--backup request not found, for example.
  * It would be useful to know what is causing the error (check allowed? send? reply?)

A little more research shows Node's base `Error` class probably has most of what I want (`code` ([kind of error](https://nodejs.org/api/errors.html#nodejs-error-codes)), `message` (text passed), `stack` (stack trace)).

I want `name` (`this.constructor.name`) if it isn't already there. One example I liked had a `cause` for validation errors--`message` has a description of the error and `cause` is the value that caused the error.

Stemmler defines `UseCaseError` that extends an interface, so loses the stack trace. While the whole trace may not be useful, I do want the name of the function that created the error. If I use an `Error` subtype, I can get it from the first line of `stack`. If I don't, I need to get the second line of from `new Error().stack` or use `Error.captureStackTrace(this)` to add a `stack` and get it from the second line.

```typescript
  // based on https://codeutility.org/javascript-get-current-function-name-in-strict-mode-stack-overflow/
  Error.captureStackTrace(this);                        // if I don't use a subtype of Error
  const stackLine = this.stack.split('\n')[2].trim();   // or 1 if a subtype of Error
  const fncName = stackLine.match(/at Object.([^ ]+)/)?.[1];
  // except that looks wrong, I want the word after 'at ', so would drop 'Object.'
```

[x] Find what's in errors by default
   * throw an `Error` and `console.log()` it
   * throw a `TypeError` and `console.log()` it
   * subclass `Error` to `TestError`, with a `cause`, throw and `console.log()` it

### What I learned
* The base error includes no keys and no members
* `console.log(e)` prints the error name (Error, TypeError, CustomError1, CustomError2), the message and a stack trace
* `e.toString()` returns the error name + message (no stack trace)
* `name`, `message`, and `stack` are set for all errors, but don't show up in keys or `JSON.stringify()` unless explicitly declared on the error (requires custom error)
* `code` seems to be `undefined` for tested types, may be set for `SystemError`, which defines it
* For Node errors, `e.name` returns the error's type (Error, TypeError, etc.)
* But for custom errors, `e.name` returns the supertype's name unless set in the constructor
* For all errors, `e.message` returns the message unless the constructor explicitly changes it
* Passing an object for `message` requires `JSON.stringify()` to make it readable.
  * Otherwise, only see `[object Object]` output.
  * The object appears in the default `console.log(e)` output as the message.
* `JSON.stringify()` returns empty for Node errors, the custom values (including custom parents) for custom errors

### Conclusions
* `class BaseError implements IBaseError`
* Include and set
  * `name` (identify the type of the error, set in constructor to ensure it's the right value)
  * `message` (error message, if an object, `JSON.stringify()` it)
  * `functionName` (name of the function that errored)
  * `fileName` (containing the function that errored)
  * optional `cause` (value or whatever caused the error, probably validation errors only)
  * optional `code` maybe (`name` is probably adequate)

**COMMIT: 3.1.7.2 - RESEARCH: default error behavior**

[x] Define `BaseError` based on findings and conclusions above
   * In `common/core` -- new subdirectory for classes that may span domain, adapter, etc.
   * Also moved Either, Guard, Result
   * Fixed paths elsewhere

Problem: In `BackupRequest`, the `create()` method returns a `Result<BackupRequest>.fail()`, which only accepts a string. Stemmler's His use cases return an `Either` that may be an error or a result and use `left()`. So unless I change `BackupRequest` to return an `Either` and use `left`, I need to figure out how to get `Result.fail()` to accept the error object. Or, I need to not put errors on `BackupResult` and instead put them on the use cases. If I do that, then I get tracing to the use case, not the place where the error realy happened.

In [this post](https://khalilstemmler.com/articles/enterprise-typescript-nodejs/functional-error-handling/), Stemmler's `fail()` accepted `any`. But when I try that, I have to add an `unknown` cast, which feels like a kludge.

Is that a major problem? Maybe not. I don't see a way around it without pushing `Either` into `BackupRequest`. I think that may cause more problems than it solves. Given that the only errors that make sense from `BackupRequest` are guard errors, I don't think I lose much (if anything), but keeping it at the use case level.

I'm also unsure why Stemmler, if `create()` fails, returns a `left(Result.fail())` that strips the error information to a string, but if the validations fail, he returns `left(new <error>)`. It's extra confusing because he uses `Result.fail<User>(userOrError.getErrorValue().toString())`. But `create()`'s failure result can only put a string in the error value. (I think this is because he declares `error: T | string`, so has to ensure it's a string.)

[x] Add `InvalidPropsError` and have the use case return it if `BackupRequest.create()` fails.

More research: Other folks (`practica`, `neverthrow`, `ts-results`, etc.) declare `Result<OkType, ErrorType>` and add functions to the class based on patterns in Rust, Go, Haskell, etc. Then have signatures like `ok<OkType, ErrorType>(value: OkType): Ok<OkType, ErrorType>` called as `ok(value)`. For example, `neverthrow` gives an example like `function addUserToDatabase(user: User): ResultAsync<User, Error> {}`, which returns a `Result` that is a `User` or an `Error`.

Stemmler does something like this in use cases where he returns an `Either` that is one of a set of errors or a `Result`. I kind of feel like his approach is a bit hacky though because it forces casting errors as the success type and restricts errors from the domain layer to strings. But for packages like `ts-results` and `neverthrow`, I'm not sure I want the whole thing. I do like `neverthrow`'s features to wrap third party libraries that may throw in `Result`s, but it also includes a lot of features I don't think I'll use. Also, the libraries I'm seeing are only getting dependency updates (at best).

Maybe the best answer is to extend Stemmler's `Result` to take an error type and change everything to support it. I can add functions based on other libraries if I need them. I like `Either`, so maybe build this on it. (The other libraries use `Option` because they're mimicing `Rust`. It's interesting than Rust's `Option` takes a single type and contains either a value of that type (`Some`) or `None`, almost like Stemmler's `Result`.)

Based on all this:
[x] Change `BaseError` to extend `Error` and take advantage of built in features you're duplicating
[x] Change `Result` to be more Rust-like
  * `ok()` and `err()` replace succeed/fail language
  * `isOk()` and `isErr()` as type guards
  * I can probably base on `neverthrow` if I remove methods I don't care about from the `IResult` interface.
  * See also `neverthrow`'s `utils` -- function to combine result list with all errors may be useful
[x] Get `CreateResultUseCase` to pass tests.
[x] Get other use cases to pass tests.
  * `InvalidPropsError` moved to `BaseErrors.ts` because it's common.
  * Aggregates and entities return `Result<type, errors>` now.
  * All error `message` strings are JSON with a `message` property plus any relevant data.
[x] Ensure errors namespaced

**COMMIT: 3.1.7.3 - REFACTOR: Result to Rust-like; code and tests**

[x] Add `ApplicationErrors.UnexpectedError`, wrap the `save()` call in a `try ... catch`, and `return left()` the error if the `save()` throws.
[x] Improve `IBackupRequestRepo` to return `Result<BackupRequest, ApplicationErrors.UnexpectedError>` for `save()` and `getById()`
[x] Update tests

**COMMIT: 3.1.7.4 - ADD: UnexpectedError; adjust tests**

[x] Rename `SendToInterfaceFailedError` to `SendToInterfaceError`
[x] Rename `InvalidPropsError` to `PropsError`
[x] Rename invalid status errors to `BackupRequestStatusError`

**COMMIT 3.1.7.5 - REFACTOR: Rename several errors (simpler, clearer, etc.)**

[x] Add `AdapterErrors.DatabaseError`
[x] Change repo - replace `UnexpectedError` with `DatabaseError`
[x] Update tests to look for `DatabaseError` from repo

**COMMIT: 3.1.7.6 - ADD: DatabaseError; adjust tests**

[x] Add `AdapterErrors.BackupJobServiceError`
[x] Change `BackupJobServiceAdapter` parts to use it
[x] Update tests to look for `BackupJobServiceError` from repo

**COMMIT: 3.1.7.7 - ADD: BackupJobServiceError; adjust tests**

## Enhance Guard
[x] Define `IGuardError` to include argName and message - only used in error case
[x] Return `Result<null, IGuardError>` from guards - don't need a result.value if it's okay
[x] Refactor code to work with changes

```typescript
interface IGuardError {
  argName: string,    // possibly useful
  message: string     // info about the failure so we can diagnose
}
```

**COMMIT: 3.1.7.8 - REFACTOR: Guard responses and code that uses them**